### PR TITLE
Teleport improvements

### PR DESCRIPTION
**Pull request by kontsevoy**
You can look into each commit separately, but here is the overview:
- Added support for arbitrary users for integration tests. You can create users separately from "teleport instances".
- Small Teleport commands like "status" or "version" do not look for web assets anymore.
- Integrated "testauthority" into integration testing
- Improved polling logic in a few places

### CODE REVIEW

**Comment by klizhentas** [comment]
there's a data race in tests:


```
time="0001-01-01T00:00:00Z" level=info msg="closing session channel" component=node fields=map[local:127.0.0.1:45992 remote:127.0.0.1:56993 login:jenkins teleportUser:jenkins id:6] file="srv/sshserver.go:600" 
time="0001-01-01T00:00:00Z" level=info msg="Asking Teleport to stop" file="_obj_test/helpers.go:419" 
time="0001-01-01T00:00:00Z" level=info msg="BroadcastEvent: event(TeleportExit)" file="service/supervisor.go:185" 
time="0001-01-01T00:00:00Z" level=info msg="client disconnected" component=proxy fields=map[local:127.0.0.1:41907 remote:127.0.0.1:37499 login:jenkins teleportUser:jenkins id:5] file="srv/sshserver.go:617" 
time="0001-01-01T00:00:00Z" level=info msg="closing session channel" component=proxy fields=map[remote:127.0.0.1:37499 login:jenkins teleportUser:jenkins id:5 local:127.0.0.1:41907] file="srv/sshserver.go:600" 
==================
WARNING: DATA RACE
Read by goroutine 86:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0x88
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0xa1c
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
==================
WARNING: DATA RACE
Read by goroutine 86:
  crypto/tls.(*listener).Close()
      <autogenerated>:35 +0x54
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0xa8
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  crypto/tls.Listen()
      /tmp/workdir/go/src/crypto/tls/tls.go:77 +0x177
  github.com/gravitational/teleport/lib/utils.ListenTLS()
      /gopath/src/github.com/gravitational/teleport/lib/utils/tls.go:46 +0x16f
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0x9d7
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
==================
WARNING: DATA RACE
Read by goroutine 86:
  net.(*TCPListener).Close()
      /tmp/workdir/go/src/net/tcpsock_posix.go:274 +0x49
  crypto/tls.(*listener).Close()
      <autogenerated>:35 +0x7a
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func5()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:716 +0xa8
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:429 +0x17a

Previous write by goroutine 83:
  net.ListenTCP()
      /tmp/workdir/go/src/net/tcpsock_posix.go:333 +0x4be
  net.Listen()
      /tmp/workdir/go/src/net/dial.go:395 +0x569
  crypto/tls.Listen()
      /tmp/workdir/go/src/crypto/tls/tls.go:73 +0xc2
  github.com/gravitational/teleport/lib/utils.ListenTLS()
      /gopath/src/github.com/gravitational/teleport/lib/utils/tls.go:46 +0x16f
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:673 +0x9d7
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 86 (running) created at:
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).onExit()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:431 +0x4c
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:718 +0x1527
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d

Goroutine 83 (running) created at:
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:148 +0x12b
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).Register()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:109 +0x3d6
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).RegisterFunc()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:121 +0x63
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxyEndpoint()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:685 +0x11e0
  github.com/gravitational/teleport/lib/service.(*TeleportProcess).initProxy.func1()
      /gopath/src/github.com/gravitational/teleport/lib/service/service.go:577 +0x5b4
  github.com/gravitational/teleport/lib/service.ServiceFunc.Serve()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:250 +0x36
  github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func2()
      /gopath/src/github.com/gravitational/teleport/lib/service/supervisor.go:144 +0x24d
==================
time="0001-01-01T00:00:00Z" level=info msg="[PROXY] proxy service exited" file="service/service.go:717" 
time="0001-01-01T00:00:00Z" level=info msg="proxySubsys(site=, host=127.0.0.1, port=45992) finished with result: <nil>" file="srv/sshserver.go:737" 
time="0001-01-01T00:00:00Z" level=info msg="[PROXY] web server exited" file="service/service.go:679" 
time="0001-01-01T00:00:00Z" level=info msg="[SUPERVISOR] Service 0x6b4bb0 is done (4)" file="service/supervisor.go:135" 
time="0001-01-01T00:00:00Z" level=info msg="TunClient.Close()" file="auth/tun.go:568" 
time="0001-01-01T00:00:00Z" level=info msg="[AUTH] auth service exited" file="service/service.go:415" 
time="0001-01-01T00:00:00Z" level=info msg="TunClient.Close()" file="auth/tun.go:568" 
time="0001-01-01T00:00:00Z" level=debug msg="TunDialer.Dial(tcp, stub:0)" file="auth/tun.go:605" 
time="0001-01-01T00:00:00Z" level=info msg="SSH server 127.0.0.1:42333 exit
```

File: lib/client/api.go [hunk 1]
```diff
@@ -326,7 +326,9 @@ func (tc *TeleportClient) SSH(command []string, runLocally bool) error {
 }
 
 // Join connects to the existing/active SSH session
-func (tc *TeleportClient) Join(sessionID session.ID) (err error) {
+func (tc *TeleportClient) Join(sid string) (err error) {
```
Comment (klizhentas) [inline lib/client/api.go]: session ID's are enforced to have `TimeUUID` format, so not sure why you need this?

File: lib/reversetunnel/agent.go [hunk 2]
```diff
@@ -361,6 +357,7 @@ func (a *Agent) startHeartbeat(conn ssh.Conn) {
 			errC <- err
 			return
 		}
+		hb.SendRequest("ping", false, nil)
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: why do you need this right after the request was sent?

File: lib/service/service.go [hunk 3]
```diff
@@ -394,12 +396,12 @@ func (process *TeleportProcess) initAuthService() error {
 			srv.Addr = fmt.Sprintf("%v:%v", process.Config.AdvertiseIP.String(), port)
 		}
 		for !askedToExit {
+			sleepTime := defaults.ServerHeartbeatTTL/2 + utils.RandomDuration(defaults.ServerHeartbeatTTL/10)
+			time.Sleep(sleepTime)
```
Comment (klizhentas) [inline lib/service/service.go]: what's the purpose of this change? looks like now you start with sleep, while before it was the opposite, just not sure what does it affect really

File: lib/client/api.go (hunk 1 above)
Comment (kontsevoy) [inline lib/client/api.go]: it's a re-typed string, causing a ton of typecasts everywhere else. so you're not gaining anything except making caller's life continuously difficult.

File: lib/reversetunnel/agent.go (hunk 2 above)
Comment (kontsevoy) [inline lib/reversetunnel/agent.go]: this speeds up integration tests because it shoots the first ping without waiting

File: lib/service/service.go (hunk 3 above)
Comment (kontsevoy) [inline lib/service/service.go]: see comments. it removes a constant "1 second" delay, replacing it with a flexible value from defaults (thus speeding up tests) and skips the first always-failing attempt at connecting a server which hasn't started yet.

File: lib/reversetunnel/agent.go (hunk 2 above)
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: but that's not what going on at least from what I read - in this particular case it simply sends ping twice

Comment (klizhentas) [inline lib/reversetunnel/agent.go]: see the line above - there's already such request executed, and your line adds one more

File: lib/client/api.go (hunk 1 above)
Comment (klizhentas) [inline lib/client/api.go]: it's not a simple string - it's a time UUID formatted ID, so these checks are necessary

Comment (klizhentas) [inline lib/client/api.go]: you are not supposed to pass any random strings there any ways, you'd need to use session.NewID()

File: lib/reversetunnel/agentpool.go [hunk 4]
```diff
@@ -107,10 +107,16 @@ func (m *AgentPool) FetchAndSyncAgents() error {
 func (m *AgentPool) pollAndSyncAgents() {
 	ticker := time.NewTicker(defaults.ReverseTunnelsRefreshPeriod)
 	defer ticker.Stop()
+	m.FetchAndSyncAgents()
 	for {
 		select {
 		case <-m.closeBroadcast.C:
 			m.Infof("closing")
+			m.Lock()
+			defer m.Unlock()
```
Comment (klizhentas) [inline lib/reversetunnel/agentpool.go]: AFAIK this introduces deadlock. as deferred function will be executed once function returns,not at the scope return, this means that any subsequent calls will block forever

File: lib/service/service.go (hunk 3 above)
Comment (klizhentas) [inline lib/service/service.go]: I don't see it (at least looking at the left side of this diff) - to me it views as simple code reshuffling. May be you did not push something?

Comment (klizhentas) [inline lib/service/service.go]: ah, my bad I see it now, nevermind

**Comment by kontsevoy** [comment]
ok, updated the branch with PR comments

File: lib/client/api.go (hunk 1 above)
Comment (kontsevoy) [inline lib/client/api.go]: The session ID is always provided _by a user via UI_ for `Join()` implementation. That's why it makes no sense to require `session.ID` type

**Comment by klizhentas** [comment]
lgtm
